src/mlpack/methods/ann/rbm/spike_slab_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,341 @@
+      false, false);
+
+  // visible penalty 1 * 1 => D * D(when used)
+  visiblePenalty = arma::mat(parameter.memptr() + weight.n_elem +
+      spikeBias.n_elem, 1, 1, false, false);
+
+  diagInvSlabPenalty = 1.0 / slabPenalty * arma::eye(poolSize, poolSize);
+  diagSlabPenalty = slabPenalty * arma::eye(poolSize, poolSize);
+}
+
+/**
+ * Free energy of the spike and slab variable
+ * the free energy of the ssRBM is given my
+ * $v^t$$\Delta$v - $\sum_{i=1}^N$ 
+ * $\log{ \sqrt{\frac{(-2\pi)^K}{\prod_{m=1}^{K}(\alpha_i)_m}}}$ -
+ * $\sum_{i=1}^N \log(1+\exp( b_i +
+ * \sum_{m=1}^k \frac{(v(w_i)_m^t)^2}{2(\alpha_i)_m})$
+ *
+ * @param input the visible layer
+ */ 
+inline double SpikeSlabRBMPolicy::FreeEnergy(arma::mat&& input)
+{
+  assert(input.n_rows == visibleSize);
+  assert(input.n_cols == 1);
+
+  scalarVisiblePenalty = visiblePenalty(0, 0);
+
+  double freeEnergy = 0.5 * arma::as_scalar(input.t() *
+      scalarVisiblePenalty * input);
+
+  assert(std::isfinite(std::log(2.0 * M_PI / slabPenalty)));
+  freeEnergy -= 0.5 * hiddenSize * poolSize *
+      std::log((2.0 * M_PI) / slabPenalty);
+
+  for (size_t i = 0; i < hiddenSize; i++)
+  {
+    double sum = 0;
+
+    for (size_t k = 0; k < poolSize; k++)
+    {
+      sum += arma::as_scalar(input.t() * weight.slice(i).col(k)) *
+          arma::as_scalar(input.t() * weight.slice(i).col(k)) /
+          (2.0 * slabPenalty);
+    }
+
+    freeEnergy -= SoftplusFunction::Fn(spikeBias(i) - sum);
+  }
+
+  return freeEnergy;
+}
+
+inline double SpikeSlabRBMPolicy::Evaluate(arma::mat& /*predictors*/,
+    size_t /*i*/)
+{
+  // Return 0 here since we don't have evaluate in case of persistence
+  return 0;
+}
+
+/**
+ * Gradient function calculates the gradient for the spike and
+ * slab RBM.
+ *
+ * @param input the visible input
+ * @param output the computed gradient
+ */
+inline void SpikeSlabRBMPolicy::PositivePhase(arma::mat&& input,
+    arma::mat&& gradient)
+{
+  assert(input.n_rows == visibleSize);
+  assert(input.n_cols == 1);
+  arma::cube weightGrad = arma::cube(gradient.memptr(), visibleSize, poolSize,
+      hiddenSize, false, false);
+
+  arma::mat spikeBiasGrad = arma::mat(gradient.memptr() + weightGrad.n_elem,
+      hiddenSize, 1, false, false);
+
+  arma::mat visiblePenaltyGrad = arma::mat(gradient.memptr() +
+      weightGrad.n_elem + spikeBiasGrad.n_elem, 1, 1, false, false);
+
+  SpikeMean(std::move(input), std::move(spikeMean));
+  SampleSpike(std::move(spikeMean), std::move(spikeSamples));
+  SlabMean(std::move(input), std::move(spikeSamples), std::move(slabMean));
+
+  // positive weight gradient
+  for (size_t i = 0 ; i < hiddenSize; i++)
+    weightGrad.slice(i) = input * slabMean.col(i).t() * spikeMean(i);
+
+  // positive hidden bias gradient
+  for (size_t i = 0; i < hiddenSize; i++)
+    spikeBiasGrad(i) = spikeMean(i);
+  // positive lambda bias
+  visiblePenaltyGrad = -0.5 * input.t() * input;
```

Comment:
The same comment. Did you compare these approaches? Are you sure that the matrix with equal diagonal elements is better? I think this question requires some research.